Remotion/Web/Development.WebTesting/WebTestSetUpFixtureHelper.cs

```diff
@@ -144,31 +144,37 @@ private void VerifyWebApplicationStarted (Uri webApplicationRoot, TimeSpan appli
 
       var stopwatch = Stopwatch.StartNew();
 
-#pragma warning disable SYSLIB0014
-      var webRequest = (HttpWebRequest)HttpWebRequest.Create(resolvedUri); // TODO RM-8492: Replace with HttpClient
-#pragma warning restore SYSLIB0014
-      webRequest.Method = WebRequestMethods.Http.Head;
-      webRequest.AllowAutoRedirect = true;
-      webRequest.Host = webApplicationRoot.Host;
-      webRequest.ServerCertificateValidationCallback += (sender, certificate, chain, errors) => true;
-
       HttpStatusCode statusCode = default;
       Assertion.DebugAssert(statusCode != HttpStatusCode.OK);
 
       while (statusCode != HttpStatusCode.OK)
       {
+  #pragma warning disable SYSLIB0014
+        var webRequest = (HttpWebRequest)HttpWebRequest.Create(resolvedUri); // TODO RM-8492: Replace with HttpClient
+  #pragma warning restore SYSLIB0014
+        webRequest.Method = WebRequestMethods.Http.Head;
+        webRequest.AllowAutoRedirect = true;
+        webRequest.Host = webApplicationRoot.GetComponents(UriComponents.Host | UriComponents.Port, UriFormat.UriEscaped);
+        webRequest.ServerCertificateValidationCallback += (sender, certificate, chain, errors) => true;
+
         try
         {
           var remainingTimeout = (int)(applicationPingTimeout.TotalMilliseconds - stopwatch.Elapsed.TotalMilliseconds);
-          webRequest.Timeout = Math.Max(remainingTimeout, 0);
-
-          using (var response = (HttpWebResponse)webRequest.GetResponse())
+          if (remainingTimeout > 0)
           {
-            statusCode = response.StatusCode;
+            webRequest.Timeout = remainingTimeout;
+
+            using (var response = (HttpWebResponse)webRequest.GetResponse())
+            {
+              statusCode = response.StatusCode;
+            }
           }
         }
         catch (WebException ex)
         {
+          if (ex.Response is HttpWebResponse httpWebResponse)
+            statusCode = httpWebResponse.StatusCode;
+
           CheckTimeout(webApplicationRoot, applicationPingTimeout, stopwatch, $"Failed with WebException '{ex.Message}'");
         }
 
```

Comment:
Not related to the current bugfix, but perhaps still something that should be fixed: if status code is 200 and no timeout happened, we still test for the timeout in line 181 and might trigger a timeout because the stopwatch is still running. And then we go to sleep for 500ms before we exit the operation. I think we should exit the method if the status code is successful (move the info log in line 188 into the getresponse in line 169.

thoughts?

Reply:
I think we don't even need to check the status code. If `.GetResponse()` did not throw the server should be reachable so we can just log and return instead of setting the status code. This allows `201` or `202` status codes as well. Something like:

```c#
while (true)
{
#pragma warning disable SYSLIB0014
  var webRequest = (HttpWebRequest)HttpWebRequest.Create(resolvedUri); // TODO RM-8492: Replace with HttpClient
#pragma warning restore SYSLIB0014
  webRequest.Method = WebRequestMethods.Http.Head;
  webRequest.AllowAutoRedirect = true;
  webRequest.Host = webApplicationRoot.GetComponents(UriComponents.Host | UriComponents.Port, UriFormat.UriEscaped);
   webRequest.ServerCertificateValidationCallback += (sender, certificate, chain, errors) => true;

  try
  {
    var remainingTimeout = (int)(applicationPingTimeout.TotalMilliseconds - stopwatch.Elapsed.TotalMilliseconds);
    if (remainingTimeout > 0)
    {
      webRequest.Timeout = remainingTimeout;

      using (var response = (HttpWebResponse)webRequest.GetResponse())
      {
        s_log.Info($"Verified that '{resolvedUri}' is accessible after {stopwatch.Elapsed.TotalMilliseconds:N0} ms.");
        return;
      }
    }
  }
  catch (WebException ex)
  {
    if (ex.Response is HttpWebResponse httpWebResponse)
      statusCode = httpWebResponse.StatusCode;

    CheckTimeout(webApplicationRoot, applicationPingTimeout, stopwatch, $"Failed with WebException '{ex.Message}'");
  }

  Thread.Sleep(TimeSpan.FromMilliseconds(500));
}
```

If you want we can do this change in this issue. I see no point in rolling it up again

Reply:
sounds good.

Reply:
The new example is a bit buggy yet, if timeout becomes zero/negative outside of a request, it will infinite-loop. I reopen the review to give it a final sanity check.